contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through a pluggable mechanism.
+    * Which allows discovery and monitoring of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Advanced scheduling and resource selection (solved through
+   [#782](https://github.com/Kubernetes/community/pull/782)).
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# Proposed Implementation 1
+
+## TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
+document and once the device plugin makes itself known to kubelet, kubelet
+will interact with the device through three simple functions:
+  1. A `ListDevices` function for the kubelet to Discover the devices and
+     their properties.
+  2. An `Allocate` function which is called before container creation
+  3. A `HealthCheck` function to notify Kubelet whenever a device becomes
+     unhealthy.
+
+![Process](device-plugin-overview.png)
+
+## Vendor story
+
+Kubernetes provides to vendors a mechanism called device plugins to:
+  * advertise devices.
+  * monitor devices (currently perform health checks).
+  * hook into the runtime to execute device specific instructions
+    (e.g: Clean GPU memory) and instruct Kubelet what are the steps
+    to take in order to make the device available in the container.
+
+```go
+service DevicePlugin {
+	rpc ListDevices(Empty) returns (stream Device) {}
+	rpc HealthCheck(Empty) returns (stream Device) {}
+
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+```
+
+The gRPC server that the device plugin must implement is expected to
+be advertised on a unix socket in a mounted hostPath (e.g:
+`/var/run/Kubernetes/vendor.sock`).
+
+Finally, to notify Kubelet of the existence of the device plugin,
+the vendor's device plugin will have to make a request to Kubelet's
+own gRPC server.
+Only then will kubelet start interacting with the vendor's device plugin
+through the gRPC apis.
+
+## End User story
+
+When setting up the cluster the admins knows what kind of devices are present
+on the different machines and therefore can select what devices he want to
+enable.
+
+The cluster admins knows his cluster has NVIDIA GPUs therefore he deploys
+the NVIDIA device plugin through:
+`kubectl create -f nvidia.io/device-plugin.yml`
+
+The device plugin lands on all the nodes of the cluster and if it detects that
+there are no GPUs it terminates. However, when there are GPUs it reports them
+to Kubelet and starts it's gRPC server to monitor devices and hook into the
+container creation process.
+
+Device Plugins reporting non-GPU Devices are advertised as OIRs of the shape
```

Comment:
These are not OIRs. To avoid confusion with OIRs can you instead refer to these resources as "Extended Resources" please?